### PR TITLE
OpdsCatalog::getSearchUrl()

### LINKED ISSUE #480: Provide method which builds the OPDS URL
Currently the process is:
1 - client builds the OPDS URL
2 - client makes the OPDS URL call and retrieve the OPDS feed
3 - client shares the OPDS feed with libkiwix which parse it and in returns gives the list of books as object

I want that the first steps is done by the libkiwix as this is a common step to all clients. A first version of that code should be already available in Kiwix Desktop (and therefore should be migrated to libkiwix and replaced by a call to new libkiwix method).

The JNI should be updated as Kiwix Android needs it really soon.

### PR DESCRIPTION
**Pull request by veloman-yunkan**
Fixes #480.

Must be merged after #526.

This is a **preliminary** implementation of OPDS search url generation targeting the new API introduced by the yet unmerged #492.

Note that #488 has enabled us to search by all fields via the query parameter. For example a search by category can be done in two ways:
1. (old) `search_endpoint?category=whatever`
2. (new) `search_endpoint?q=category:whatever`

Internally, the old approach is converted to an equivalent of the new one with the slight difference of not assuming partial query (which is the default mode for the `q` parameter).

Some fields (creator, publisher) are not exposed for searching in the old scheme but are available via the new scheme.

Now the question is which one of the two schemes should be utilized for the search URL generation?

Another question is what should be done about the filtering criteria `Filter::local()`, `Filter::remote()`, `Filter::remote()`? They have never been exposed through the OPDS search API.

TODOs:
- Don't ignore the `Filter::maxSize()` & `Filter::rejectTags()` filtering criteria.

### CODE REVIEW

**Comment by kelson42** [comment]
@veloman-yunkan Thx, do you have a kiwix-desktop related PR, so I can just test if it works (no regression) from user perspective?

**Comment by veloman-yunkan** [comment]
> @veloman-yunkan Thx, do you have a kiwix-desktop related PR, so I can just test if it works (no regression) from user perspective?

No. It won't work until the new OPDS catalog API is merged and goes live on `library.kiwix.org` (or you will have to test with a custom-built `kiwix-serve` incorporating #492).

**Comment by kelson42** [comment]
@veloman-yunkan Really hard to me then to verify it works fine from a user persective. Will let @mgautierfr do the review.

**Review by mgautierfr — CHANGES_REQUESTED** [review]
> Now the question is which one of the two schemes should be utilized for the search URL generation?

I prefer the old one. Mainly because it is easier to read.

**Comment by veloman-yunkan** [comment]
@kelson42 I will finalize this PR after #492 is merged (which introduces new catalog API that is assumed here).

File: src/opds_catalog.cpp
Comment (veloman-yunkan) [reply src/opds_catalog.cpp:71]: I tried it but didn't like the result. `getSearchUrl()` starts looking like a plagiarist that delegates all work to a ghost writer but receives all the credit. The current division of work between `getSearchUrl()` and `buildSearchString()` is more balanced.

**Comment by kelson42** [comment]
@veloman-yunkan #492 has been merge, I guess nothing stops us anymore to finish this PR?

**Comment by veloman-yunkan** [comment]
@kelson42 This PR is finished and pending review by @mgautierfr

**Comment by mgautierfr** [comment]
Nothing to say on the code.

I'm a bit less sure about the generated endpoint.
As the catalog V2 is not totally finalized, we may change the endpoints and we agree to not use them until we finish it (at least until we are confident enough to tell we can start us it).
If `getSearchUrl` use the v2 and we do a release with `getSearchUrl` then user will start to use it and start to use the v2 catalog. We have to be careful with this.

I will not block this PR for this, but I may block the next release however :slightly_smiling_face:

**Comment by mgautierfr** [comment]
rebase/fixup on master.

This PR is open for two long now. As we agree on the code we can merge it.
We will discuss again this v2 endpoint when we will do the release.

**Review by mgautierfr — APPROVED** [review]

**Pull request merged**
